networkx/algorithms/d_separation.py

```diff
@@ -291,6 +301,11 @@ def minimal_d_separator(G, u, v):
     # perform BFS on the graph from 'x' to mark
     Z_dprime = _bfs_with_marks(moral_G, u, Z_prime)
     Z = _bfs_with_marks(moral_G, v, Z_dprime)
+
+    # check if the set Z actually separates 'u' and 'v'
+    if not d_separated(G, {u}, {v}, Z):
+        Z = None
```

Comment:
Is there a case when Z is non-empty and yet `d_separated(G, {u}, {v}, Z)` returns `False`? 

If so, how can we avoid returning None?
If not, let's check whether `Z` is an empty set before calling `d_separated`. 
```python
if Z or d_separated(G, {u}, {v}, Z):
    return Z
return None
```

Reply:
I think that example, if correct, shows that our current code has an error in it. 
Why is the implemented algorithm not providing a correct answer? Do we need to change everything?

I am learning this causal analysis material -- I'm no expert. But I believe the papers that present these algorithms are correct. Are they really wrong?  Do we implement an incorrect version of the algorithms? Do you see my confusion?

I'm beginning to think that this entire module needs a complete overhaul. Hopefully not...

Reply:
Sorry, perhaps I misspoke and was not clear, but I believe the paper and our implementation to be both correct.

The way the algorithm functions is that it will propose a set Z as a candidate minimal d-separating set. If at least one d-separating set exists, then Z will be the minimal d-separating set (I can probably hunt down the reference if it comes to that), and the algorithm returns Z. If not, then Z clearly cannot be a d-separating set, and thus not a minimal d-separating set. Thus the added d-separation check ensures that the candidate minimal set is in fact a separating set in the first place.

The example I have provided above is such a case where there is no d-separating set between A and B. The algorithm proposed C as the candidate separating set, but in the d-separation check it will surface that C does not d-separate A and B. The correct answer to return should be None (which the algorithm currently does according to the tests). 

Hopefully this explanation clarifies the situation.

Reply:
For a  reference see page 13 of this article: http://arxiv.org/abs/1803.00116. The relevant function is FINDMINSEP where we consider I=\emptyset and R={all vertices in graph} (this article considers mixed graphs which contain directed, bidirected and undirected edges, but DAGs are a subclass of such a graph). Note that in a DAG the anterior (Ant, in the article) is equivalent to ancestors. In this function, line 7 has a check that the candidate set Z is in fact a minimal d-separating set.

I note that the original reference in the code https://ftp.cs.ucla.edu/pub/stat_ser/r254.pdf fails to take into account the situation if there is no d-separating set at all, which is an error to the best of my knowledge.

Reply:
So our module has many functions that are, at best, incomplete and at worst wrong. The `is_minimal_d_separator` function only checks minimal -- which we've decided is an error. And the `minimal_d_separator` function returns sets that are not separators -- which seems like an even more error-like error.

Can we reconsider the entire API for this set of functions?  The function names should reflect what the function does. And we should figure out a way to express (and possibly separate) the notions of d_separable and minimal_d_separator so that it is clear what we are computing, checking or not computing, checking.

Thoughts?

Reply:
What about renaming 

minimal_d_separator to generate_minimal_d_separators to make it explicit that this is generating such sets?

And change is_minimal_d_separator to minimal_d_separated since this is exactly the d_separated function except it also checks for minimality. This keeps the naming in line with the current d_separated function. 

I think functionally everything should still stay as is because they are all independently useful. Perhaps make documentation clearer on what is checked or generated vs what is not. 

Wdyt?

Reply:
I think perhaps `find_minimal_d_separator` would be a natural naming, since it leaves open the possibility that we fail to find such a minimal d-separator. I don't think there is anything one can do about the functionality of  `minimal_d_separator` since a user can specify an arbitrary DiGraph and there is no guarantee that a minimal d-separating set will exist in general. 

+1 on documentation, we can definitely improve the clarity there